src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs

```diff
@@ -65,51 +70,69 @@ internal partial class DiagnosticService : IDiagnosticService
+                        continue;
+                    }
 
-            _eventQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
-            {
-                var updatedArgsCollection = UpdateDataMap(source, argsCollection);
-                if (updatedArgsCollection.IsEmpty)
-                {
-                    // there is no change, nothing to raise events for.
-                    return;
+                    // don't create event listener if it haven't created yet. if there is a diagnostic to remove
+                    // listener should have already created since all events are done in the serialized queue
+                    ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.args), (source, args: argsBuilder.ToImmutableAndClear()));
                 }
+            }
 
-                ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.updatedArgsCollection), (source, updatedArgsCollection));
-            }, CancellationToken.None);
+            return ValueTaskFactory.CompletedTask;
         }
 
-        private void RaiseDiagnosticsCleared(IDiagnosticUpdateSource source)
+        private void RaiseDiagnosticsUpdated(IDiagnosticUpdateSource source, ImmutableArray<DiagnosticsUpdatedArgs> argsCollection)
         {
-            var ev = _eventMap.GetEventHandlers<EventHandler<ImmutableArray<DiagnosticsUpdatedArgs>>>(DiagnosticsUpdatedEventName);
-
-            _eventQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
+            if (argsCollection.IsEmpty)
             {
-                using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
+                // No work to do. Avoid adding an empty args collection to _eventQueue because that is used to represent
+                // an operation from RaiseDiagnosticsCleared.
+                return;
+            }
 
-                if (!ClearDiagnosticsReportedBySource(source, ref argsBuilder.AsRef()))
+            Workspace? previousWorkspace = null;
+            foreach (var args in argsCollection)
+            {
+                if (args.Workspace != previousWorkspace)
                 {
-                    // there is no change, nothing to raise events for.
-                    return;
+                    _eventListenerTracker.EnsureEventListener(args.Workspace, this);
+                    previousWorkspace = args.Workspace;
                 }
+            }
+
+            _eventQueue.AddWork((source, argsCollection));
+        }
 
-                // don't create event listener if it haven't created yet. if there is a diagnostic to remove
-                // listener should have already created since all events are done in the serialized queue
-                ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.args), (source, args: argsBuilder.ToImmutableAndClear()));
-            }, CancellationToken.None);
+        private void RaiseDiagnosticsCleared(IDiagnosticUpdateSource source)
```

Comment:
i'm not really understanding who is calling this.

Reply:
The caller is abstracted a few layers. There is an event hookup here:

https://github.com/dotnet/roslyn/blob/3ab3ca743a0a702a1dd925c40a969dc9c660dd5b/src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService_UpdateSourceRegistrationService.cs#L24